### PR TITLE
Fix/talk2scholars buglist

### PR DESCRIPTION
**Pull request by ansh-info**
# For authors

image

## Description

Added Human in the loop for write operations using the [documentation](https://langchain-ai.github.io/langgraph/concepts/human_in_the_loop/)

The key improvements in this compliant version:

1. **Proper Implementation of `interrupt()`**: The zotero_review_tool now properly uses the `interrupt()` function as described in the LangGraph documentation for human-in-the-loop. This attempts to pause execution of the graph and wait for user input.

2. **Fallback Mechanism**: I still maintain a fallback for when the interrupt functionality isn't available, where it asks the user to respond with "Yes" and passes that confirmation to the save tool.

3. **State Management**: The tools properly maintain state about what's been reviewed and approved, following the pattern shown in the documentation.

4. **Multiple Approval Paths**: I handle both interrupt-based approval and message-based approval, giving flexibility to different UI environments.

These changes should bring our implementation much more in line with the LangGraph documentation. The key LangGraph human-in-the-loop concepts (interrupt, Command, state management for breakpoints) are now properly implemented.

This approach merges the simplicity of current implementation with the proper use of LangGraph features described in the documentation.

### How I Fixed Bug 1 (Zotero Write Fails if Zotero Read Hasn't Been Called):

- The root cause was that the write tool depended on state data populated by the read tool
- I fixed this by:
  - Adding the `fetch_papers_for_save()` function to properly check state
  - Adding clear error messages when no papers are found
  - Removing the direct dependency on `zotero_read` being in the state

### How I Fixed Bug 2 (Incorrect Collection Target):

- The root cause was improper collection path matching that used partial matches
- I fixed this by:
  - Implementing `find_or_create_collection()` that does proper hierarchical matching
  - Testing parent-child relationships at each level of the path
  - Providing more informative error messages with available collections
  - Using exact collection name matching instead of partial matching

## Fixes # (issue) Mention the issue number.
#142  Bug fix (1,2) -> (already solved -solved 3, 4, 5, 6 in pr #143 )

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [x] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [x] No new tests added (Please explain the rationale in this case)

## Checklist
- [ ] My code follows the style guidelines mentioned in the Code/DevOps guides
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [ ] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: aiagents4pharma/talk2scholars/state/state_talk2scholars.py [hunk 1]
```diff
@@ -54,6 +54,8 @@ class Talk2Scholars(AgentState):
         papers (Dict[str, Any]): Stores the research papers retrieved from the agent's queries.
         multi_papers (Dict[str, Any]): Stores multiple recommended papers from various sources.
         zotero_read (Dict[str, Any]): Stores the papers retrieved from Zotero.
+        approved_zotero_save (Dict[str, Any]): Stores the approval status and collection
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/state/state_talk2scholars.py:57]: I'd recommend using a more appropriate name, e.g.: zotero_write_approval_status.

File: aiagents4pharma/talk2scholars/agents/zotero_agent.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 from langgraph.checkpoint.memory import MemorySaver
 from ..state.state_talk2scholars import Talk2Scholars
 from ..tools.zotero.zotero_read import zotero_search_tool
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/agents/zotero_agent.py:16]: I recommend using the same name while importing for readability purposes.

File: aiagents4pharma/talk2scholars/agents/zotero_agent.py [hunk 3]
```diff
@@ -14,6 +14,7 @@
...
+from ..tools.zotero.zotero_review_tool import zotero_review_tool
 from ..tools.zotero.zotero_write import zotero_save_tool
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/agents/zotero_agent.py:18]: I recommend using the same name while importing for readability purposes.

File: aiagents4pharma/talk2scholars/agents/zotero_agent.py [hunk 4]
```diff
@@ -14,6 +14,7 @@
 from langgraph.checkpoint.memory import MemorySaver
 from ..state.state_talk2scholars import Talk2Scholars
 from ..tools.zotero.zotero_read import zotero_search_tool
+from ..tools.zotero.zotero_review_tool import zotero_review_tool
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/agents/zotero_agent.py:17]: If you plan to add the prefix "_tool" to the tool name, I recommend doing the same for all the tools.

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_review_tool.py
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/tools/zotero/zotero_review_tool.py:1]: I wonder if it's necessary to create an additional tool for validation when the [validation method](https://langchain-ai.github.io/langgraph/concepts/human_in_the_loop/#validating-human-input) in LangGraph can already handle it within zotero_write.

**Review by gurdeep330 — CHANGES_REQUESTED** [review]
Bravo @ansh-info  Thank you for fixing the bugs and implementing human-in-the-loop. I do have some comments, mainly regarding the alternative methods available for implementing human-in-the-loop. Let me know your thoughts.

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 5]
```diff
@@ -30,33 +32,141 @@ class ZoteroSaveInput(BaseModel):
...
+                            tool_call_id=tool_call_id,
+                        )
+                    ],
+                    "approved_zotero_save": {"approved": False},
+                }
+            )
+
+    # Now proceed with normal checks
+    if not approval_info:
+        return Command(
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:104]: Sure, Thank you for pointing out

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 6]
```diff
@@ -235,15 +283,26 @@ def zotero_save_tool(
...
 
+    # pylint: disable=broad-exception-caught
     # Save items to Zotero
     try:
         response = zot.create_items(zotero_items)
         logger.info("Papers successfully saved to Zotero: %s", response)
     except Exception as e:
         logger.error("Error saving to Zotero: %s", str(e))
-        raise RuntimeError(f"Error saving papers to Zotero: {str(e)}") from e
+        return Command(
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:293]: Im sorry, I reverted back to it

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 7]
```diff
@@ -30,33 +32,141 @@ class ZoteroSaveInput(BaseModel):
...
     """
+    # First check if we have approval from the interrupt mechanism
+    approval_info = state.get("approved_zotero_save", {})
+
+    # If we don't have explicit approval but have papers_reviewed state and user confirmation
+    if not approval_info.get("approved", False) and approval_info.get(
+        "papers_reviewed", False
+    ):
+        # Check if user has confirmed in this non-interrupt workflow
+        if user_confirmation.lower() in [
```
[the remark this answers is not in the document]
Comment (dmccloskey) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:78]: This solution will be problematic. If moving to an LLM is not possible at this time, I would recommend only allowing for a single confirmation e.g., "Yes" or "No", adding the single confirmation as the only option when pinging the user for their approval, and continuously pinging the user until one of the confirmations is provided.

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 8]
```diff
@@ -30,33 +32,141 @@ class ZoteroSaveInput(BaseModel):
...
+                    ToolMessage(
+                        content="Save operation was rejected by the user.",
+                        tool_call_id=tool_call_id,
+                    )
+                ],
+            }
+        )
+
+    # If papers were reviewed but not approved and no confirmation provided
+    if (
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:131]: Is it possible to change the implementation so that this point is never reached (i.e., review but no approval)?

**Review by dmccloskey — CHANGES_REQUESTED** [review]
This is a really nice start to implementing human in the loop for zotero writes 👍. My major concerns are the following:

1. How can the user verify that the write operation occurred as the user "expected"? Is there a way to display the updated zotero collection to the user?

2. What happens when the user would like to "undo" a write operation? Is there a way to rollback an update to the zotero database?

3. What happens if the user really messes up, and they find themselves in a situation where a simple "undo" would not be help? Is there a way to backup the zotero collection before modifying it during a session?

File: aiagents4pharma/talk2scholars/agents/zotero_agent.py (hunk 2 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/zotero_agent.py:16]: Sure, I made the changes

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py (hunk 8 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:123]: Yes, I initially added it as a fallback, but I removed it later as I read more about interrupts and developed a better understanding

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py]: Thank you for the feedback, @dmccloskey and @gurdeep330 -> this particular logic was for a fallback, i.e: if interrupt fails -> and now I have removed it(initially added it as a fallback, but I removed it later as I read more about interrupts and developed a better understanding),

and I am working toward integrating structured output as of now

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_review_tool.py
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_review_tool.py:1]: Yes, you are correct, I tried this initially this approach and made a utility but, I was facing graph recursion and The documentation confirms that a while loop with interrupt calls should work but even after using a while True loop with a single interrupt() call inside it, it never break

Hence I moved to a tool, since it was easy to implement, I am still working on, hopefully I would try to save it in this pr, if not I would come to this in the subsequent pr's


```python
    # using a while loop
    # with an interrupt call inside
    try:
        approval = None
        while True:
            # Single interrupt call that might be repeated if validation fails
            approval = interrupt(prompt)

            logger.info("Received approval response: %s", approval)

            # If valid, break out of the loop
            if is_valid_response(approval):
                break
 ```
 
 
image

File: aiagents4pharma/talk2scholars/configs/agents/talk2scholars/zotero_agent/default.yaml [hunk 9]
```diff
@@ -1,19 +1,36 @@
...
+     - AND the user_confirmation parameter containing the user's "Yes" message
 
-  WORKFLOW STEPS
-  1. When user requests papers, use `zotero_search_tool` to find papers
-  2. Use `display_results` tool to display the response
-  3. Use `query_results` tool to query over the selected paper only when the user asks to
-  4. Use `retrieve_semantic_scholar_paper_id` to get the semantic scholar id of a paper title for the papers from zotero library
-  5. Use `zotero_write` to save the papers to users zotero library under collections and call `display_results` only after you recive
-     the save was successfull
+  Your available tools are:
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/agents/talk2scholars/zotero_agent/default.yaml:20]: Yes, its correct -> I was just following the pattern from s2's prompt, I have reduced the prompt now, but if required I can do further changes

File: aiagents4pharma/talk2scholars/state/state_talk2scholars.py (hunk 1 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/state/state_talk2scholars.py]: Sure

File: aiagents4pharma/talk2scholars/state/state_talk2scholars.py
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/state/state_talk2scholars.py:1]: Thank you, for now I have hashed it -> meanwhile I am working of implementic a custom reducer which would append instead of replacing the existing dict, I would want to add it in the subsequent pr's

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py]: Hi @dmccloskey and @gurdeep330, I have added `with_structured_output` in zotero_review using Literal.

https://github.com/user-attachments/assets/eddec185-8829-4143-9672-c83a4e9c7acf

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_review_tool.py
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_review_tool.py:1]: The current solution is working well and avoids the issues we previously encountered (graph recursion and repeated printing). However, to align fully with the human-in-the-loop documentation, we need to refine the implementation(for this pr I will still be using the new tool zotero_review). I've created issue #150  to track these improvements, and I'll collaborate with @gurdeep330 on this in future updates.

**Comment by ansh-info** [comment]
> This is a really nice start to implementing human in the loop for zotero writes 👍. My major concerns are the following:
> 
> 1. How can the user verify that the write operation occurred as the user "expected"? Is there a way to display the updated zotero collection to the user?
> 2. What happens when the user would like to "undo" a write operation? Is there a way to rollback an update to the zotero database?
> 3. What happens if the user really messes up, and they find themselves in a situation where a simple "undo" would not be help? Is there a way to backup the zotero collection before modifying it during a session?

Thank you for the review :

To answers you questions:

1. Currently, we save the papers from the `last_displayed_papers` key immediately after the save operation, without returning the saved papers to a dedicated key (such as `zotero_write`). If needed, we can create this key, which would help in cases where users want to save only a specific paper from `last_displayed_papers`. After the write operation, we could store the saved paper in the `zotero_write` key and later use our `display_results` tool to provide proper feedback.

2. At the moment, we do not support an undo operation. I will need to review the pyzotero documentation for this, but I believe we could create a snapshot of the existing library when calling the `zotero_read` or `zotero_write` tools. We could store that snapshot in a key and later perform a rollback operation if necessary. I'm not certain if this is the only solution, but it is a potential approach that I will investigate further. Thanks for raising this important question, especially since we are dealing with users' sensitive data.(I hope this answers the 3rd question as well) 😀

**Comment by dmccloskey** [comment]
> 1. Currently, we save the papers from the `last_displayed_papers` key immediately after the save operation, without returning the saved papers to a dedicated key (such as `zotero_write`). If needed, we can create this key, which would help in cases where users want to save only a specific paper from `last_displayed_papers`. After the write operation, we could store the saved paper in the `zotero_write` key and later use our `display_results` tool to provide proper feedback.

Gotcha. My remark was more towards providing the user with greater visibility into their current Zotero library.  As you show in the video, the user can go to the Zotero UI to confirm changes. However, it would be beneficial for the adoption of T2S if the user does not have to know or understand how Zotero works in the background.

> 2. At the moment, we do not support an undo operation. I will need to review the pyzotero documentation for this, but I believe we could create a snapshot of the existing library when calling the `zotero_read` or `zotero_write` tools. We could store that snapshot in a key and later perform a rollback operation if necessary. I'm not certain if this is the only solution, but it is a potential approach that I will investigate further. Thanks for raising this important question, especially since we are dealing with users' sensitive data.(I hope this answers the 3rd question as well) 😀

Creating a snapshot of items, collections, and tags would be a good first step. It looks like we can take advantage of the "Trash" https://www.zotero.org/support/collections_and_tags for any delete operations that the user would like to undo. I think we could also be creative with collections and tags for tracking operations and enabling "undo"/"redo".

**Review by dmccloskey — APPROVED** [review]
My comments appear to be resolved: nice work 💪. Our discussion in regard to "undo/redo" functionality and better visibility into a user's collection can be handled in subsequent PRs.

**Comment by ansh-info** [comment]
> My comments appear to be resolved: nice work 💪. Our discussion in regard to "undo/redo" functionality and better visibility into a user's collection can be handled in subsequent PRs.

Thank you, to cover this I have created an issue #152 
and I will read about the link provided

**Review by gurdeep330 — CHANGES_REQUESTED** [review]
@ansh-info thanks for addressing the requests. The pylint tests have failed. Do you mind fixing them, please? Thx!

**Comment by ansh-info** [comment]
> @ansh-info thanks for addressing the requests. The pylint tests have failed. Do you mind fixing them, please? Thx!

Hi, It was related to the hashed log statement for state, I will look for a work around other than hashing.

**Review by gurdeep330 — APPROVED** [review]
